### PR TITLE
Add support for NGINX VOD Module

### PR DESCRIPTION
**Pull request by hunterjm**
This adds support for viewing recordings as a "video on demand". It can be tested by loading the stream URL into [this demo site](http://players.akamai.com/players/hlsjs) or VLC.

The VOD stream format currently follows the recording directory format - i.e.: http://localhost:5000/vod/2021-05/17/20/{camera}/master.m3u8

This first version is obviously limited to an hour's worth of recordings based on the current recording file structure. We can expand it later by modifying the `vod` function in `http.py`. Regardless, The Home Assistant integration can be updated to use this new URL to allow Media Browser recordings of entire hour long segments without having to use FFMPEG to combine the videos. The Web UI can also use this along with something like `videojs` to build a recording playback page. I left the `/recordings` directive in tact for backwards compatibility until things switch over.

### CODE REVIEW

File: frigate/app.py
Comment (hunterjm) [inline frigate/app.py:322]: This doesn't necessarily belong as a part of this PR, but in testing I noticed some exceptions being raised on `ctrl+C` so this handles it.

File: frigate/log.py
Comment (hunterjm) [inline frigate/log.py:38]: Similar to above, an exception was being raised here on shutdown, so handling it here.

File: frigate/http.py
Comment (hunterjm) [inline frigate/http.py:453]: This builds a JSON object to be consumed by `nginx-vod-module`. Currently we are just using the existing recording path, but can expand this to add additional filters or functionality later.

File: docker/Dockerfile.nginx
Comment (hunterjm) [inline docker/Dockerfile.nginx:1]: We need to build NGINX from source to include support for aio files and multiple architectures. `nginx-vod-module` has a repo we could add, but it is only built for `amd64`.

File: docker/Dockerfile.base
Comment (hunterjm) [inline docker/Dockerfile.base:22]: removing `nginx` and `libnginx-mod-rtmp` as they are built-in to our compiled version of NGINX along with `nginx-vod-module`.

**Review by hunterjm — COMMENTED** [review]
Hope these comments/notes are helpful as you look through this.

**Comment by hunterjm** [comment]
Will update docs before merging if this looks like a good path forward to you.

File: frigate/http.py
Comment (blakeblackshear) [inline frigate/http.py:466]: How long does it take to do this dynamically? It could take a while if using a network drive.

**Review by blakeblackshear — APPROVED** [review]

File: frigate/http.py
Comment (hunterjm) [reply frigate/http.py:466]: It’s two seconds on my dev machine, but it’s also pretty beefy and the files were local. I almost just hardcoded the 60 seconds here, but know that the command (and thus the time) can be changed in configuration.

I saw the use of ffprobe elsewhere for clips, but since OpenCV was already imported and [this stackoverflow answer](https://stackoverflow.com/a/61572332) said it was faster, I just used it. Granted I did not do any verification myself. It’s probably worth a shot testing this on a Pi with video on a network drive to see.

**Comment by mitchross** [comment]
Just out of curiosity, what benefits would this have for the 'clips' feature?

**Comment by hunterjm** [comment]
> Just out of curiosity, what benefits would this have for the 'clips' feature?

This doesn't really have any benefit for clips as they are non-continuous and already are the length you expect them to be. This is just a way to combine the 60 second recording files together in an hour long video to make viewing the 24/7 recording easier.

What we could do, eventually, is in the front-end automatically add markers to the timeline for the continuous recording at the timestamps that events occurred to be able to seek to them with relative ease.

**Comment by stephengolub** [comment]
I literally just built a small script to generate a downloadable mp4 file to facilitate large chunks of time. I'm wondering if instead of `path`, you take in `camera` and then have query parameters that are date/time objects...

Then you just glob the files and do a greater than/less than check. You could default to the last 60 minutes

**Comment by hunterjm** [comment]
> I literally just built a small script to generate a downloadable mp4 file to facilitate large chunks of time. I'm wondering if instead of `path`, you take in `camera` and then have query parameters that are date/time objects...
> 
> Then you just glob the files and do a greater than/less than check. You could default to the last 60 minutes.
> 
> This is my relevant bash code:
> 
> ```shell
> file_list=$(ls **/**/**/$CAMERA/*.mp4)                                                                                    
>                                                                                                                           
> real_list=()                                                                                                              
>                                                                                                                           
> for file in $file_list; do                                                                                                
>     if [ "$file" \> "$START_FILE" ] && [ "$file" \< "$END_FILE" ]; then                                                   
>         real_list+=("file '$PWD/$file'")                                                                                  
>     fi                                                                                                                    
> done 
> ```

I honestly think that's the next step, but wanted to keep the initial implementation simple (easier to review just the relevant VOD module code) and expand the functionality of filters/etc in a follow-up PR.

**Comment by stephengolub** [comment]
If you take a look at the PR I opened against your branch, I illustrated what I meant (I haven't tested it fully yet), but, tbh, I don't think it's too much more complicated.

https://github.com/hunterjm/frigate/pull/1

**Comment by hunterjm** [comment]
> If you take a look at the PR I opened against your branch, I illustrated what I meant (I haven't tested it fully yet), but, tbh, I don't think it's too much more complicated.
> 
> [hunterjm#1](https://github.com/hunterjm/frigate/pull/1)

After some initial testing, query string parameters do not get passed on from the VOD module. We can do some path based filtering like `camera/start/end` But another potential issue is the amount of time it takes to recursively glob the entire recording directory and iterate over the 43,200 files that would exist with the default retention period. I am having trouble finding a fast solution for just the latest 24 hours. I’m happy to research further and put it in a follow-up PR, but don’t want to block this one because of it.

**Comment by stephengolub** [comment]
I was afraid of that. Everyone knows you shouldn’t use the directory tree as a database.

Might be best to table it like you said and work on optimizing how we look up available recordings and clips

**Comment by hunterjm** [comment]
Did some testing on the `cv.VideoCapture` time for a full hour's worth of recordings (60) on my dev box:

Glob: 0.000562
Sort: 0.000004
CV Read Total: 2.192811
CV Read Avg: 0.036547

versus FFProbe that is used to verify clips and recordings:

Glob: 0.000504
Sort: 0.000005
FFProbe Read Total: 2.946326
FFProbe Read Avg: 0.049105

Granted it is a Ryzen 3900x so it may take longer on a Raspberry Pi, but it's encouraging that the OpenCV method is not only faster than a ffprobe sub process, but is less lines of code as well.

The only other option would be to default duration to 60, as that is what it will be the vast majority of time. I figured if it didn't take too long to be certain though, it's better for cases where recordings ended early or a user modified the recording segment time in their configuration.

I'm happy with this PR now - the next commit updating docs should be the last.

File: frigate/http.py
Comment (hunterjm) [reply frigate/http.py:466]: Verified timing in comment below.

**Comment by hunterjm** [comment]
@blakeblackshear - could you run the nginx make command to push the images to docker hub? That would make it so others can successfully build frigate.

**Comment by blakeblackshear** [comment]
My buildx command is failing to build the armv7 image. Did you do some set up to get it to work?

**Comment by hunterjm** [comment]
No, it built just fine for me.

**Comment by hunterjm** [comment]
Built and pushed for all 3 architectures here: https://hub.docker.com/r/hunterjm/frigate-nginx/tags?page=1&ordering=last_updated

**Comment by blakeblackshear** [comment]
Got it working with [these steps](https://github.com/docker/buildx/issues/495#issuecomment-778896842). Mine are [pushed](https://hub.docker.com/repository/docker/blakeblackshear/frigate-nginx).

**Pull request merged**
